mllib/src/main/scala/org/apache/spark/ml/tree/impl/DecisionTreeMetadata.scala

```diff
@@ -184,15 +185,15 @@ private[spark] object DecisionTreeMetadata extends Logging {
       case _ => featureSubsetStrategy
     }
 
-    val isIntRegex = "^([1-9]\\d*)$".r
-    val isFractionRegex = "^(0?\\.\\d*[1-9]\\d*|1\\.0+)$".r
     val numFeaturesPerNode: Int = _featureSubsetStrategy match {
       case "all" => numFeatures
       case "sqrt" => math.sqrt(numFeatures).ceil.toInt
       case "log2" => math.max(1, (math.log(numFeatures) / math.log(2)).ceil.toInt)
       case "onethird" => (numFeatures / 3.0).ceil.toInt
-      case isIntRegex(number) => if (BigInt(number) > numFeatures) numFeatures else number.toInt
-      case isFractionRegex(fraction) => (fraction.toDouble * numFeatures).ceil.toInt
+      case RandomForestParams.integerFeatureSubsetStrategy(number) =>
```

Comment:
I think we should simplify the checks.

~~~scala
import scala.util.Try

val numFeaturesPerNode: Int = _featureSubsetStrategy match {
   ...
   case _ =>
     Try(_featureSubsetStrategy.toDouble).toOption match {
       case Some(value) =>
         if (value <= 0.0) {
           throw new IllegalArgumentException(...)
         } else if (value < 1.0) {
           math.ceil(value * numFeatures)
         } else {
           math.ceil(value)
         }
       case None =>
         throw new IllegalArgumentException(...)
     }
}
~~~

See my comment below for the validator.